In [ ]:
!pip install gensim

In [ ]:
import gensim
from gensim.models import Word2Vec,keyedvectors

In [ ]:
import gensim.downloader as api
wv = api.load('word2vec-google-news-300')
vec_king = wv['king']

[==================================================] 100.0% 1662.8/1662.8MB downloaded


In [ ]:
vec_king.shape #its a 300 dimensional vector

(300,)

In [ ]:
vec_king

array([ 1.25976562e-01,  2.97851562e-02,  8.60595703e-03,  1.39648438e-01,
       -2.56347656e-02, -3.61328125e-02,  1.11816406e-01, -1.98242188e-01,
        5.12695312e-02,  3.63281250e-01, -2.42187500e-01, -3.02734375e-01,
       -1.77734375e-01, -2.49023438e-02, -1.67968750e-01, -1.69921875e-01,
        3.46679688e-02,  5.21850586e-03,  4.63867188e-02,  1.28906250e-01,
        1.36718750e-01,  1.12792969e-01,  5.95703125e-02,  1.36718750e-01,
        1.01074219e-01, -1.76757812e-01, -2.51953125e-01,  5.98144531e-02,
        3.41796875e-01, -3.11279297e-02,  1.04492188e-01,  6.17675781e-02,
        1.24511719e-01,  4.00390625e-01, -3.22265625e-01,  8.39843750e-02,
        3.90625000e-02,  5.85937500e-03,  7.03125000e-02,  1.72851562e-01,
        1.38671875e-01, -2.31445312e-01,  2.83203125e-01,  1.42578125e-01,
        3.41796875e-01, -2.39257812e-02, -1.09863281e-01,  3.32031250e-02,
       -5.46875000e-02,  1.53198242e-02, -1.62109375e-01,  1.58203125e-01,
       -2.59765625e-01,  

In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv('/content/spam.csv',encoding='latin-1')

In [ ]:
df = df.rename(columns={'v1':'label','v2':'text'})

In [ ]:
df = df.drop(['Unnamed: 2','Unnamed: 3','Unnamed: 4'],axis=1)

In [ ]:
df = df.dropna()

In [ ]:
df

label                                               text
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...
...    ...                                                ...
5567  spam  This is the 2nd time we have tried 2 contact u...
5568   ham              Will Ì_ b going to esplanade fr home?
5569   ham  Pity, * was in mood for that. So...any other s...
5570   ham  The guy did some bitching but I acted like i'd...
5571   ham                         Rofl. Its true to its name

[5572 rows x 2 columns]

In [ ]:
import re
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [ ]:
from nltk.stem import WordNetLemmatizer
lemmatizer = WordNetLemmatizer()
nltk.download('wordnet')


[nltk_data] Downloading package wordnet to /root/nltk_data...


True

In [ ]:
corpus = []
for i in range (0,len(df)):

  review = re.sub('[^a-zA-Z]',' ',df['text'][i]) #it says that apart from a to z and A to Z,if you find any special character,replace with blank.
  review = review.lower()
  review = review.split()
  review = [lemmatizer.lemmatize(word) for word in review ]
  review = ' '.join(review)
  corpus.append(review)

In [ ]:
[[i,j,k] for i,j,k in zip(list(map(len,corpus)),corpus,df['text']) if i<1]

[[0, '', '645'], [0, '', ':) '], [0, '', ':-) :-)']]

these are the 3 records which got removed from x_new during preprocessing .bcoz in the above code we said that apart from a to z and A to Z, remove everything.so it removed.

In [ ]:
corpus

['go until jurong point crazy available only in bugis n great world la e buffet cine there got amore wat',
 'ok lar joking wif u oni',
 'free entry in a wkly comp to win fa cup final tkts st may text fa to to receive entry question std txt rate t c s apply over s',
 'u dun say so early hor u c already then say',
 'nah i don t think he go to usf he life around here though',
 'freemsg hey there darling it s been week s now and no word back i d like some fun you up for it still tb ok xxx std chgs to send to rcv',
 'even my brother is not like to speak with me they treat me like aid patent',
 'a per your request melle melle oru minnaminunginte nurungu vettam ha been set a your callertune for all caller press to copy your friend callertune',
 'winner a a valued network customer you have been selected to receivea prize reward to claim call claim code kl valid hour only',
 'had your mobile month or more u r entitled to update to the latest colour mobile with camera for free call the mobile up

In [ ]:
from nltk import sent_tokenize
from gensim.utils import simple_preprocess #it converts a document into a list of lowercase tokens,ignoring that are too short or too long.

In [ ]:
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [ ]:
words=[]
for sent in corpus:
  sent_token = sent_tokenize(sent)
  for sent in sent_token:
    words.append(simple_preprocess(sent)) #here for every sentence, i am getting list of words

In [ ]:
words

[['go',
  'until',
  'jurong',
  'point',
  'crazy',
  'available',
  'only',
  'in',
  'bugis',
  'great',
  'world',
  'la',
  'buffet',
  'cine',
  'there',
  'got',
  'amore',
  'wat'],
 ['ok', 'lar', 'joking', 'wif', 'oni'],
 ['free',
  'entry',
  'in',
  'wkly',
  'comp',
  'to',
  'win',
  'fa',
  'cup',
  'final',
  'tkts',
  'st',
  'may',
  'text',
  'fa',
  'to',
  'to',
  'receive',
  'entry',
  'question',
  'std',
  'txt',
  'rate',
  'apply',
  'over'],
 ['dun', 'say', 'so', 'early', 'hor', 'already', 'then', 'say'],
 ['nah',
  'don',
  'think',
  'he',
  'go',
  'to',
  'usf',
  'he',
  'life',
  'around',
  'here',
  'though'],
 ['freemsg',
  'hey',
  'there',
  'darling',
  'it',
  'been',
  'week',
  'now',
  'and',
  'no',
  'word',
  'back',
  'like',
  'some',
  'fun',
  'you',
  'up',
  'for',
  'it',
  'still',
  'tb',
  'ok',
  'xxx',
  'std',
  'chgs',
  'to',
  'send',
  'to',
  'rcv'],
 ['even',
  'my',
  'brother',
  'is',
  'not',
  'like',
  'to',
  'spea

In [ ]:
import gensim

In [ ]:
#lets train word2vec from scratch
model = gensim.models.Word2Vec(words

)

In [ ]:
#to get all the vocabulary
model.wv.index_to_key

['you',
 'to',
 'the',
 'and',
 'it',
 'in',
 'is',
 'me',
 'my',
 'for',
 'your',
 'call',
 'of',
 'that',
 'have',
 'on',
 'now',
 'are',
 'can',
 'so',
 'but',
 'not',
 'or',
 'we',
 'do',
 'get',
 'at',
 'be',
 'if',
 'will',
 'ur',
 'with',
 'no',
 'just',
 'this',
 'gt',
 'lt',
 'go',
 'how',
 'up',
 'when',
 'ok',
 'day',
 'what',
 'free',
 'from',
 'all',
 'out',
 'know',
 'll',
 'come',
 'like',
 'time',
 'good',
 'then',
 'am',
 'got',
 'wa',
 'there',
 'he',
 'text',
 'only',
 'love',
 'want',
 'send',
 'txt',
 'need',
 'one',
 'today',
 'going',
 'by',
 'home',
 'don',
 'about',
 'stop',
 'she',
 'lor',
 'sorry',
 'see',
 'still',
 'mobile',
 'take',
 'back',
 'da',
 'reply',
 'our',
 'think',
 'dont',
 'tell',
 'week',
 'phone',
 'hi',
 'new',
 'later',
 'they',
 'any',
 'her',
 'pls',
 'ha',
 'please',
 'co',
 'did',
 'msg',
 'been',
 'min',
 'an',
 'some',
 'dear',
 'make',
 'here',
 'night',
 'message',
 'who',
 'say',
 'well',
 're',
 'where',
 'thing',
 'much',
 'grea

In [ ]:
model.corpus_count

5569

In [ ]:
model.epochs #default is 5 but more no.of epochs,more better

5

In [ ]:
model.wv.similar_by_word('good')

[('day', 0.9992080926895142),
 ('hope', 0.9989867806434631),
 ('happy', 0.9988439679145813),
 ('great', 0.9988434910774231),
 ('my', 0.9988214373588562),
 ('thing', 0.9987605214118958),
 ('well', 0.9987408518791199),
 ('morning', 0.9986963868141174),
 ('why', 0.9986485838890076),
 ('all', 0.9986413717269897)]

In [ ]:
model.wv['good'].shape

(100,)

In [ ]:
model.wv['good']

array([-0.2674552 ,  0.2170276 ,  0.33270767,  0.23403825,  0.20262781,
       -0.6597683 ,  0.31045866,  0.87983584, -0.44252333, -0.37567887,
       -0.21154283, -0.63776815, -0.12956254,  0.20356737,  0.1923092 ,
       -0.22804621,  0.05838401, -0.37660927, -0.00178112, -0.70859003,
        0.21913552,  0.3634409 ,  0.26685154, -0.35280412, -0.02639521,
        0.1056322 , -0.36515924, -0.28084412, -0.39337286,  0.06995106,
        0.37325534, -0.11224447,  0.22990449, -0.1146429 , -0.24041551,
        0.377496  ,  0.13804518, -0.33410633,  0.0169181 , -0.6289896 ,
        0.06287998, -0.20956822, -0.21054935, -0.05847096,  0.32895553,
       -0.12112987, -0.31942376, -0.10316022,  0.06629138,  0.21167868,
        0.06334062, -0.21746305, -0.1740879 ,  0.01613747, -0.0585198 ,
        0.13070622,  0.31441724,  0.09335919, -0.4131536 ,  0.2978362 ,
       -0.01231713,  0.05942227,  0.11789227, -0.11737309, -0.4613684 ,
        0.34142235,  0.12253115,  0.2511095 , -0.35558283,  0.41

In [ ]:
words[0]

['go',
 'until',
 'jurong',
 'point',
 'crazy',
 'available',
 'only',
 'in',
 'bugis',
 'great',
 'world',
 'la',
 'buffet',
 'cine',
 'there',
 'got',
 'amore',
 'wat']

here if we apply word2vec, we will get 100 dimensional vectors for each and every word in the sentence. but we want for every sentence,a 100 dimensional vector.

In [ ]:
#here we use avgword2vec
def avg_word2vec(doc):
  return np.mean([model.wv[word] for word in doc if word in model.wv.index_to_key],axis=0)

In [ ]:
!pip install tqdm

In [ ]:
from tqdm import tqdm

In [ ]:
#apply for entire sentences
import numpy as np
x=[]
for i in tqdm(range(len(words))):
  x.append(avg_word2vec(words[i]))

  8%|▊         | 447/5569 [00:00<00:01, 4468.59it/s]/usr/local/lib/python3.10/dist-packages/numpy/core/fromnumeric.py:3504: RuntimeWarning: Mean of empty slice.
  return _methods._mean(a, axis=axis, dtype=dtype,
/usr/local/lib/python3.10/dist-packages/numpy/core/_methods.py:129: RuntimeWarning: invalid value encountered in scalar divide
  ret = ret.dtype.type(ret / rcount)
100%|██████████| 5569/5569 [00:01<00:00, 4623.29it/s]


In [ ]:
x

[array([-0.16777952,  0.16607483,  0.24033792,  0.17039609,  0.1266405 ,
        -0.4508774 ,  0.19047493,  0.6281548 , -0.31924206, -0.24459504,
        -0.16976257, -0.42384815, -0.09030929,  0.14200002,  0.12777473,
        -0.16041476,  0.04285717, -0.2664454 ,  0.01172522, -0.50635546,
         0.14907597,  0.24089827,  0.19410606, -0.23575623, -0.02275748,
         0.08710346, -0.25516474, -0.17859961, -0.29751527,  0.06334291,
         0.24552569, -0.07775753,  0.14631046, -0.10324211, -0.15654983,
         0.2795177 ,  0.08304559, -0.2438452 , -0.02195249, -0.42973801,
         0.02117183, -0.15548787, -0.15957032, -0.03539006,  0.22850157,
        -0.09452123, -0.19697271, -0.08782291,  0.05455529,  0.1513436 ,
         0.04069976, -0.18127266, -0.14408989,  0.0238136 , -0.04419085,
         0.08198378,  0.21166089,  0.05749891, -0.30270252,  0.19243495,
        -0.01293856,  0.02139098,  0.0911672 , -0.0787204 , -0.33421725,
         0.24792394,  0.09827873,  0.17933029, -0.2

In [ ]:
len(x)

5569

In [ ]:
#independent features
x_new = np.array(x)

ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (5569,) + inhomogeneous part.

In [ ]:
# Inspect the shape of each element
for i, arr in enumerate(x):
    print(f"Array {i} shape: {arr.shape}")

# Define the target shape
target_shape = (100,)  # Assuming you want each array to have 100 elements

# Fix the shapes of the arrays
x_fixed_shape = [np.resize(arr, target_shape) if arr.shape != target_shape else arr for arr in x]

# Verify the shapes after fixing
for i, arr in enumerate(x_fixed_shape):
    print(f"Fixed Array {i} shape: {arr.shape}")

# Convert the list of fixed arrays to a NumPy array
x_new = np.array(x_fixed_shape)
print(f"Combined array shape: {x_new.shape}")

Streaming output truncated to the last 5000 lines.
Fixed Array 570 shape: (100,)
Fixed Array 571 shape: (100,)
Fixed Array 572 shape: (100,)
Fixed Array 573 shape: (100,)
Fixed Array 574 shape: (100,)
Fixed Array 575 shape: (100,)
Fixed Array 576 shape: (100,)
Fixed Array 577 shape: (100,)
Fixed Array 578 shape: (100,)
Fixed Array 579 shape: (100,)
Fixed Array 580 shape: (100,)
Fixed Array 581 shape: (100,)
Fixed Array 582 shape: (100,)
Fixed Array 583 shape: (100,)
Fixed Array 584 shape: (100,)
Fixed Array 585 shape: (100,)
Fixed Array 586 shape: (100,)
Fixed Array 587 shape: (100,)
Fixed Array 588 shape: (100,)
Fixed Array 589 shape: (100,)
Fixed Array 590 shape: (100,)
Fixed Array 591 shape: (100,)
Fixed Array 592 shape: (100,)
Fixed Array 593 shape: (100,)
Fixed Array 594 shape: (100,)
Fixed Array 595 shape: (100,)
Fixed Array 596 shape: (100,)
Fixed Array 597 shape: (100,)
Fixed Array 598 shape: (100,)
Fixed Array 599 shape: (100,)
Fixed Array 600 shape: (100,)
Fixed Array 601 sha

In [ ]:
x_new

array([[-0.16777952,  0.16607483,  0.24033792, ..., -0.20345739,
         0.09605334, -0.07288297],
       [-0.15183944,  0.14985746,  0.20757277, ..., -0.1907135 ,
         0.0825293 , -0.0590667 ],
       [-0.17125064,  0.16505645,  0.29861319, ..., -0.2021655 ,
         0.06394827, -0.07808802],
       ...,
       [-0.23381397,  0.22186604,  0.31019431, ..., -0.27603856,
         0.14523183, -0.09393486],
       [-0.21082728,  0.19538458,  0.29442984, ..., -0.24982275,
         0.11456934, -0.0870055 ],
       [-0.23755059,  0.20536713,  0.31050941, ..., -0.2669197 ,
         0.12260242, -0.0864471 ]])

In [ ]:
x_new.shape

(5569, 100)

In [ ]:
x_new[0].shape

(100,)

In [ ]:
#dependent feature
#covert label column to numerical variables
y = df[list(map(lambda x:len(x)>0,corpus))]
y = pd.get_dummies(y['label'])
y = y.iloc[:,0].values

In [ ]:
y.shape

(5569,)

In [ ]:
y

array([ True,  True, False, ...,  True,  True,  True])

In [ ]:
y_new = y.astype(int)

In [ ]:
y_new

array([1, 1, 0, ..., 1, 1, 1])

In [ ]:
df.shape

(5572, 2)

if you see here our independent column x_new has 5569 rows,whereas output has 5572 rows. its a problem.i.e, after preprocessing we have lost 3 records

In [ ]:
#our final independent features
dq = pd.DataFrame(x_new)

In [ ]:
dq.shape

(5569, 100)

In [ ]:
dq.head()

0         1         2         3         4         5         6   \
0 -0.167780  0.166075  0.240338  0.170396  0.126640 -0.450877  0.190475   
1 -0.151839  0.149857  0.207573  0.142662  0.109359 -0.392072  0.162040   
2 -0.171251  0.165056  0.298613  0.215731  0.142585 -0.483618  0.161822   
3 -0.241086  0.221279  0.303804  0.219389  0.166887 -0.606637  0.257252   
4 -0.205685  0.182243  0.269334  0.183811  0.152834 -0.511832  0.220412   

         7         8         9   ...        90        91        92        93  \
0  0.628155 -0.319242 -0.244595  ...  0.344821  0.209268  0.008077  0.119462   
1  0.540431 -0.271284 -0.207375  ...  0.303764  0.165100  0.001471  0.109790   
2  0.683351 -0.338047 -0.265680  ...  0.369414  0.222681  0.001462  0.138097   
3  0.830457 -0.423641 -0.324268  ...  0.456574  0.258389  0.004092  0.171160   
4  0.702204 -0.362652 -0.280871  ...  0.392105  0.224733  0.006886  0.143740   

         94        95        96        97        98        99  
0  0.505969  0.296788  0.086323 -0.203457  0.096053 -0.072883  
1  0.425856  0.249764  0.064378 -0.190713  0.082529 -0.059067  
2  0.506071  0.322399  0.085807 -0.202165  0.063948 -0.078088  
3  0.674243  0.389404  0.106591 -0.290692  0.135459 -0.080522  
4  0.569203  0.334185  0.089448 -0.250855  0.114528 -0.074203  

[5 rows x 100 columns]

In [ ]:
dq['output']=y_new

In [ ]:
dq.head()

0         1         2         3         4         5         6  \
0 -0.167780  0.166075  0.240338  0.170396  0.126640 -0.450877  0.190475   
1 -0.151839  0.149857  0.207573  0.142662  0.109359 -0.392072  0.162040   
2 -0.171251  0.165056  0.298613  0.215731  0.142585 -0.483618  0.161822   
3 -0.241086  0.221279  0.303804  0.219389  0.166887 -0.606637  0.257252   
4 -0.205685  0.182243  0.269334  0.183811  0.152834 -0.511832  0.220412   

          7         8         9  ...        91        92        93        94  \
0  0.628155 -0.319242 -0.244595  ...  0.209268  0.008077  0.119462  0.505969   
1  0.540431 -0.271284 -0.207375  ...  0.165100  0.001471  0.109790  0.425856   
2  0.683351 -0.338047 -0.265680  ...  0.222681  0.001462  0.138097  0.506071   
3  0.830457 -0.423641 -0.324268  ...  0.258389  0.004092  0.171160  0.674243   
4  0.702204 -0.362652 -0.280871  ...  0.224733  0.006886  0.143740  0.569203   

         95        96        97        98        99  output  
0  0.296788  0.086323 -0.203457  0.096053 -0.072883       1  
1  0.249764  0.064378 -0.190713  0.082529 -0.059067       1  
2  0.322399  0.085807 -0.202165  0.063948 -0.078088       0  
3  0.389404  0.106591 -0.290692  0.135459 -0.080522       1  
4  0.334185  0.089448 -0.250855  0.114528 -0.074203       1  

[5 rows x 101 columns]

In [ ]:
dq.dropna(inplace=True)

In [ ]:
dq.isnull().sum()

0         0
1         0
2         0
3         0
4         0
         ..
96        0
97        0
98        0
99        0
output    0
Length: 101, dtype: int64

In [ ]:
#independent feature
X = dq.iloc[:,:-1]

In [ ]:
X.isnull().sum()

0     0
1     0
2     0
3     0
4     0
     ..
95    0
96    0
97    0
98    0
99    0
Length: 100, dtype: int64

In [ ]:
y_new = dq['output']

In [ ]:
#train test split
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y_new,test_size=0.2,random_state=2)

In [ ]:
X_train.head()

0         1         2         3         4         5         6   \
13   -0.214006  0.189152  0.293743  0.205401  0.166410 -0.549328  0.228220   
2713 -0.211545  0.194815  0.288621  0.204728  0.156166 -0.543856  0.227168   
3756 -0.160785  0.148914  0.228654  0.159949  0.122417 -0.421323  0.165388   
4837 -0.171896  0.147073  0.219712  0.149715  0.133947 -0.429932  0.181737   
370  -0.233536  0.190833  0.281267  0.189962  0.177477 -0.558057  0.240706   

            7         8         9   ...        90        91        92  \
13    0.756930 -0.388909 -0.306988  ...  0.413751  0.251457  0.001436   
2713  0.751462 -0.387802 -0.295895  ...  0.413804  0.248344  0.003236   
3756  0.580678 -0.294870 -0.225802  ...  0.319597  0.188590 -0.000548   
4837  0.587171 -0.302178 -0.233197  ...  0.325578  0.189102  0.003738   
370   0.755150 -0.393701 -0.315861  ...  0.427806  0.238783 -0.004740   

            93        94        95        96        97        98        99  
13    0.147909  0.601908  0.361180  0.097586 -0.260191  0.128335 -0.076455  
2713  0.144957  0.606255  0.357770  0.104913 -0.249965  0.121530 -0.079383  
3756  0.116043  0.458917  0.273907  0.076877 -0.196669  0.090000 -0.059638  
4837  0.116407  0.476590  0.279581  0.080520 -0.213769  0.099132 -0.057191  
370   0.152451  0.611591  0.368155  0.080222 -0.300330  0.142062 -0.064355  

[5 rows x 100 columns]

In [ ]:
y_train

13      1
2713    1
3756    1
4837    1
370     1
       ..
3340    1
1101    1
2517    1
3611    1
2579    0
Name: output, Length: 4445, dtype: int64

In [ ]:
from sklearn.ensemble import RandomForestClassifier
classifier=RandomForestClassifier()

In [ ]:
classifier.fit(X_train,y_train)

RandomForestClassifier()

In [ ]:
y_pred=classifier.predict(X_test)

In [ ]:
from sklearn.metrics import accuracy_score,classification_report,confusion_matrix
score=accuracy_score(y_test,y_pred)

In [ ]:
print(score)

0.9640287769784173


In [ ]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.87      0.82      0.84       132
           1       0.98      0.98      0.98       980

    accuracy                           0.96      1112
   macro avg       0.92      0.90      0.91      1112
weighted avg       0.96      0.96      0.96      1112

